skilloracle/skilloracle/skilloracle/skilloracle.py

```diff
@@ -0,0 +1,193 @@
+        where .close doesn't really close and .connect doesn't through
+        see: https://github.com/requests/requests/issues/1882
+
+        I'm probably overlooking something simple but we favor the
+        echo, netcat strategy below for now. Note that w/o shell,
+        netcat also hangs like the socket approach. This
+        approach requires less resource management and debugging.
+        """
+        ret = None
+        echo = subprocess.Popen(('echo', content), stdout=subprocess.PIPE)
+        try:
+            nc = subprocess.check_output(('netcat', host, str(port)),
+                                         stdin=echo.stdout,
+                                         shell=True,
+                                         timeout=10)
+        except subprocess.CalledProcessError:
+            ret = False # nothing listening on the port
+        else:
+            ret = True
+
+        return ret
+
+    def check_socket(self, host=None, port=None):
+        ret = False
+        host = host
+        if host == None:
+            host = '127.0.0.1' # need some kind of a host to check
+
+        self.sendrecv(host, port, "1")
+        time.sleep(0.5) # wait for a replay
+        ret = True
+
+        return ret
+
+    def kill(self, name='vw'):
+        ret = subprocess.call(shlex.split('killall vw'))
+        return ret == 0
+
+    def PUT(self, label, name, context):
+        response = None
+
+        # todo: I think this function expects a string array
+        name = escape_vw_string(name)
+        context = escape_vw_string(context)
+
+        if label:
+            label = escape_vw_string(label)
+            labelled_example = "{label} |{context_namespace} {context} \
+                                        |{name_namespace} {name}".\
+                    format(label=label,
+                           context_namespace="context",
+                           context=context,
+                           name_namespace="name",
+                           name=name)
+        else:
+            labelled_example = "|{context_namespace} {context} \
+                                |{name_namespace} {name}".\
+                    format(context_namespace="context",
+                           context=context,
+                           name_namespace="name",
+                           name=name)
+
+            self.oracle.sendline(labelled_example)
+            response = self.oracle._recvline()
+
+        return response
+
+    def GET(self):
+        # note: __get_redis() shoudl be cleaned up
+        # just return importance, key/skill candidate
+        response = self.__get_redis()
+        return response
+
+    def __setup_redis(self):
+        raise NotImplementedError
+
+    def __get_redis(self):
+        # see: https://groups.google.com/forum/#!topic/redis-db/ur9U8o-Sko0
+        response = None
+        pipe = self.redis_db.pipeline()# runs w/in multi/exec, atomic, on Redis
+
+        pipe.zrange(self.SKILL_CANDIDATES,
```

Comment:
I'm curious what these three commands are retrieving. You're running three commands, and then in the endpoint you grab three pieces of data (candidate, context, importance). Am I correct in assuming that each of these commands is producing one of those pieces of data?

Some comments inline describing the Redis data structures being used and how they're used would be helpful. You may have this in some documentation elsewhere, but inlining some would make the code more clear.